postgresql/resource_postgresql_grant.go

```diff
@@ -269,8 +271,12 @@ WHERE grantee = $2
 	if err := txn.QueryRow(query, dbName, roleOID).Scan(&privileges); err != nil {
 		return fmt.Errorf("could not read privileges for database %s: %w", dbName, err)
 	}
-
-	d.Set("privileges", pgArrayToSet(privileges))
+	granted := pgArrayToSet(privileges)
+	wanted := d.Get("privileges").(*schema.Set)
+	equal := arePrivilegesEqual(granted, wanted, d)
+	if !equal {
+		return d.Set("privileges", wanted)
```

Comment:
`granted` no ? as this function read the state of the database. Otherwise there will never be any diff as you set in the state what's in the state already.
```suggestion
		return d.Set("privileges", granted)
```

Reply:
I am not quite sure about that. We check whats on the DB...

 if the state on the DB does not equal whats requested via terraform, we want to set the privileges that are requested. `granted` is what we already have.

Reply:
the only thing that is weird to me right now is the fact, that I fetch the required privileges from the resourcedata and then put it back there again. this call is quite superflous but still won't hurt.

Reply:
> we want to set the privileges that are requested. granted is what we already have.

Indeed, but that's not exactly how this function works.

The goal of a resource read function (in this case `resourcePostgreSQLGrantRead`) is to refresh the Terraform state with the current status of what's in Postgres. 

Then Terraform will compare the resource as it's defined in your code with how it is in the state (so in Postgres).
So the goal of this line is to set the privileges as it is in the dabase so it will be compared with what's defined in you resource definition.

> Since i tested the changes with a database locally and had great success with the former implementation

It's because you haven't tested all the use cases (if I dare :sweat_smile: )

You think you fully fix that because there's no more diff in `terraform plan` compared to before, but actually there will now never be any diff because you say that the current state in the database is what's defined in your code, so Terraform will always say that everything is already correct.

Easy to demonstrate with this simple code:

```hcl
resource "postgresql_database" "test" { 
  name = "test_db"
}

resource "postgresql_role" "test" {
  name = "test"
}

resource "postgresql_grant" "db_all" {
  database    = postgresql_database.test.name
  role        = postgresql_role.test.name
  object_type = "database"
  privileges  = ["ALL"]
}
```
that you can apply:
```
terraform apply
```

Then, you manually remove a permission in Postgres:
```
$ psql -c "revoke connect ON database test_db from test "
REVOKE
```

And next `terraform plan` will not show any diff when it should say that it needs to fix the permissions.


If I rebuild my provider with the suggestion I did above (set `granted` instead of wanted), if I manually remove a permission from the database, I will have this plan:

```
$ terraform plan
[...]
Terraform will perform the following actions:

  # postgresql_grant.db_all must be replaced
-/+ resource "postgresql_grant" "db_all" {
      ~ id                = "test_test_db_database" -> (known after apply)
      ~ privileges        = [ # forces replacement
          - "CREATE",
          - "TEMPORARY",
          + "ALL",
        ]
        # (4 unchanged attributes hidden)
    }
```
And once applied, thanks to your fix, there will not be any diff in the next plan.

Reply:
.. that sounds reasonable 😅 i adjusted that with [54fc593](https://github.com/cyrilgdn/terraform-provider-postgresql/pull/339/commits/54fc593f610136ed3d41b95dbe9f826de0a49bf6) and provided an acceptance test for it. would be cool if you could check again @cyrilgdn 

thanks in advance